File: sdk/react-native/dazzle-react-native/src/edge/openAICompatibleClient.ts
// Copyright 2026 Ivan Aliaga <ivan.aliaga@urp.edu.pe>
// SPDX-License-Identifier: Apache-2.0
//
// OpenAI-compatible HTTP client — pure JS over fetch + line-based SSE.
// Works with any endpoint that speaks `POST /v1/chat/completions` with
// the OpenAI JSON shape (OpenAI, HF Router, Ollama, vLLM, Groq, …).

import { Completion, Delta, Message, Role } from '../agent/message';
import { LLMClient } from '../agent/llmClient';
import { ToolDeclaration, serializeSchema } from '../agent/tool';

export interface OpenAICompatibleOptions {
  baseURL: string;
  model: string;
  apiKey?: string;
  extraHeaders?: Record<string, string>;
  temperature?: number;
  maxTokens?: number;
  fetchImpl?: typeof fetch;
}

export class OpenAICompatibleClient implements LLMClient {
  readonly modelId: string;
  private readonly baseURL: string;
  private readonly model: string;
  private readonly apiKey?: string;
  private readonly extraHeaders: Record<string, string>;
  private readonly temperature?: number;
  private readonly maxTokens?: number;
  private readonly fetchImpl: typeof fetch;

  constructor(opts: OpenAICompatibleOptions) {
    this.baseURL = opts.baseURL.replace(/\/+$/, '');
    this.model = opts.model;
    this.modelId = opts.model;
    this.apiKey = opts.apiKey;
    this.extraHeaders = opts.extraHeaders ?? {};
    this.temperature = opts.temperature;
    this.maxTokens = opts.maxTokens;
    this.fetchImpl = opts.fetchImpl ?? fetch;
  }

  async complete(opts: {
    messages: Message[]; tools?: ToolDeclaration[];
  }): Promise<Completion> {
    const body = this.buildBody(opts.messages, opts.tools ?? [], false);
    const resp = await this.fetchImpl(`${this.baseURL}/chat/completions`, {
      method: 'POST',
      headers: this.headers(),
      body: JSON.stringify(body),
    });
    // Read body once. Some RN fetch polyfills set resp.ok = false
    // for chunked-encoding 200s — check status directly instead.
    const raw = await resp.text();
    if (resp.status < 200 || resp.status >= 300) {
      throw new Error(`OpenAI HTTP ${resp.status}: ${raw}`);
    }
    // HF Router (Groq) ignores `stream:false` and always emits SSE.
    // Detect by the leading `data: ` and fold the chunks into a
    // single completion. Plain JSON path is the OpenAI / Ollama
    // happy path.
    const json = raw.trimStart().startsWith('data:')
        ? foldSseChunks(raw)
        : JSON.parse(raw);
    const choice = json.choices?.[0];
    const msg = choice?.message ?? choice?.delta ?? {};
    if (msg.tool_calls?.length) {
      return {
        type: 'toolCalls',
        message: {
          role: 'assistant',
          content: msg.content ?? '',
          toolCalls: msg.tool_calls.map((c: any) => ({
            id: c.id,
            name: c.function?.name ?? '',
            arguments: c.function?.arguments ?? '{}',
          })),
        },
      };
    }
    return {
      type: 'text',
      message: { role: 'assistant', content: msg.content ?? '' },
    };
  }

  async *stream(opts: {
    messages: Message[]; tools?: ToolDeclaration[];
  }): AsyncIterable<Delta> {
    const body = this.buildBody(opts.messages, opts.tools ?? [], true);
    const dbg = !!(globalThis as any).DAZZLE_DEBUG_LLM_BODY;
    if (dbg) {
      // eslint-disable-next-line no-console
      console.log('[DAZZLE_DEBUG] LLM body =', JSON.stringify(body));
    }
    const resp = await this.fetchImpl(`${this.baseURL}/chat/completions`, {
      method: 'POST',
      headers: this.headers(),
      body: JSON.stringify(body),
    });
    if (dbg) {
      // eslint-disable-next-line no-console
      console.log('[DAZZLE_DEBUG] resp.status =', resp.status,
                  'body=', resp.body ? 'present' : 'null',
                  'getReader=', typeof (resp.body as any)?.getReader);
    }
    if (resp.status < 200 || resp.status >= 300) {
      throw new Error(`OpenAI HTTP ${resp.status}: ${await resp.text()}`);
    }
    // React Native's fetch polyfill returns `resp.body === null` even
    // for 200s — it doesn't ship a streaming ReadableStream. Fall back
    // to buffered text + chunk-as-SSE so we still get incremental
    // semantics (the agent doesn't care if all deltas land in one tick
    // for a 256-token reply).
    const activeCalls = new Map<number, { id: string; name: string }>();
    if (!resp.body || typeof (resp.body as any).getReader !== 'function') {
      const raw = await resp.text();
      if (dbg) {
        // eslint-disable-next-line no-console
        console.log('[DAZZLE_DEBUG] buffered SSE bytes =', raw.length,
                    'head=', raw.slice(0, 220));
      }
      yield* parseSseChunks(raw, activeCalls);
      return;
    }
    const reader = (resp.body as any).getReader();
    const decoder = new TextDecoder();
    let buffer = '';
    while (true) {
      const { done, value } = await reader.read();
      if (done) break;
      buffer += decoder.decode(value, { stream: true });
      let idx: number;
      while ((idx = buffer.indexOf('\n')) >= 0) {
        const line = buffer.slice(0, idx).trim();
        buffer = buffer.slice(idx + 1);
        if (!line || !line.startsWith('data:')) continue;
        const payload = line.slice(5).trim();
        if (payload === '[DONE]') { yield { type: 'end' }; return; }
        let data: any;
        try { data = JSON.parse(payload); } catch { continue; }
        const delta = data.choices?.[0]?.delta ?? {};
        if (typeof delta.content === 'string' && delta.content.length) {
          yield { type: 'text', chunk: delta.content };
        }
        if (Array.isArray(delta.tool_calls)) {
          for (const c of delta.tool_calls) {
            const idx2 = c.index ?? 0;
            if (c.id || c.function?.name) {
              const prev = activeCalls.get(idx2);
              const id   = c.id ?? prev?.id ?? `call_${idx2}`;
              const name = c.function?.name ?? prev?.name ?? '';
              if (!prev || prev.id !== id) {
                activeCalls.set(idx2, { id, name });
                yield { type: 'toolCallStart', id, name };
              }
            }
            if (c.function?.arguments) {
              const prev = activeCalls.get(idx2);
              if (prev) {
                yield { type: 'toolCallArgs', id: prev.id, chunk: c.function.arguments };
              }
            }
          }
        }
      }
    }
    yield { type: 'end' };
  }

  async close(): Promise<void> {}

  // ── helpers ───────────────────────────────────────────────────────

  private headers(): Record<string, string> {
    const h: Record<string, string> = {
      'Content-Type': 'application/json',
      ...this.extraHeaders,
    };
    if (this.apiKey) h.Authorization = `Bearer ${this.apiKey}`;
    return h;
  }

  private buildBody(
      messages: Message[], tools: ToolDeclaration[], stream: boolean): any {
    const body: any = {
      model: this.model,
      messages: messages.map((m) => messageToJson(m)),
      stream,
    };
    if (this.temperature !== undefined) body.temperature = this.temperature;
    if (this.maxTokens !== undefined) body.max_tokens = this.maxTokens;
    if (tools.length) {
      body.tools = tools.map((t) => ({
        type: 'function',
        function: {
          name: t.name,
          description: t.description,
          parameters: JSON.parse(serializeSchema(t.parameters)),
        },
      }));
    }
    return body;
  }
}

/// Yield Delta events from a buffered SSE response — same shape as
/// the streaming reader path, just consumed in one shot. Used when
/// `resp.body` isn't a streaming ReadableStream (RN whatwg-fetch).
async function* parseSseChunks(
    raw: string,
    activeCalls: Map<number, { id: string; name: string }>): AsyncIterable<Delta> {
  const dbg = !!(globalThis as any).DAZZLE_DEBUG_LLM_BODY;
  let nDataLines = 0;
  let nText = 0;
  let nToolCalls = 0;
  for (const line of raw.split('\n')) {
    const t = line.trim();
    if (!t.startsWith('data:')) continue;
    const payload = t.slice(5).trim();
    if (!payload) continue;
    if (payload === '[DONE]') {
      if (dbg) {
        // eslint-disable-next-line no-console
        console.log('[DAZZLE_DEBUG] SSE done — dataLines=', nDataLines,
                    'text=', nText, 'toolCalls=', nToolCalls);
      }
      yield { type: 'end' } as Delta;
      return;
    }
    nDataLines++;
    let chunk: any;
    try { chunk = JSON.parse(payload); } catch (e) {
      if (dbg) {
        // eslint-disable-next-line no-console
        console.log('[DAZZLE_DEBUG] SSE parse failed for', payload.slice(0, 120));
      }
      continue;
    }
    const delta = chunk.choices?.[0]?.delta ?? {};
    if (typeof delta.content === 'string' && delta.content.length) {
      nText += delta.content.length;
      yield { type: 'text', chunk: delta.content } as Delta;
    }
    if (Array.isArray(delta.tool_calls)) {
      for (const tc of delta.tool_calls) {
        const idx = tc.index ?? 0;
        const prev = activeCalls.get(idx);
        const id   = tc.id ?? prev?.id ?? `call_${idx}`;
        const name = tc.function?.name ?? prev?.name ?? '';
        if (!prev || prev.id !== id) {
          activeCalls.set(idx, { id, name });
          nToolCalls++;
          yield { type: 'toolCallStart', id, name } as Delta;
        }
        if (tc.function?.arguments) {
          const cur = activeCalls.get(idx);
          if (cur) yield {
            type: 'toolCallArgs', id: cur.id, chunk: tc.function.arguments,
          } as Delta;
        }
      }
    }
  }
  if (dbg) {
    // eslint-disable-next-line no-console
    console.log('[DAZZLE_DEBUG] SSE end-no-DONE — dataLines=', nDataLines,
                'text=', nText, 'toolCalls=', nToolCalls);
  }
  yield { type: 'end' } as Delta;
}

/// Reassemble a non-streaming completion from a stream of SSE
/// `data: {...}` chunks. Some providers (notably HF Router → Groq)
/// always send chunked output regardless of `stream:false`. Each
/// chunk is a partial OpenAI-shaped completion in `choices[].delta`;
/// we concatenate `delta.content` and merge the last `tool_calls`.
function foldSseChunks(raw: string): any {
  const out: any = { choices: [{ delta: { content: '' } }] };
  for (const line of raw.split('\n')) {
    const t = line.trim();
    if (!t.startsWith('data:')) continue;
    const payload = t.slice(5).trim();
    if (!payload || payload === '[DONE]') continue;
    let chunk: any;
    try { chunk = JSON.parse(payload); } catch { continue; }
    const delta = chunk.choices?.[0]?.delta ?? {};
    if (delta.role) out.choices[0].delta.role = delta.role;
    if (typeof delta.content === 'string') {
      out.choices[0].delta.content += delta.content;
    }
    if (delta.tool_calls?.length) {
      out.choices[0].delta.tool_calls = delta.tool_calls;
    }
  }
  return out;
}

function messageToJson(m: Message): any {
  const role: Role = m.role;
  const out: any = { role, content: m.content };
  if (m.toolCalls?.length) {
    out.tool_calls = m.toolCalls.map((c) => ({
      id: c.id,
      type: 'function',
      function: { name: c.name, arguments: c.arguments },
    }));
  }
  if (m.toolCallId) out.tool_call_id = m.toolCallId;
  return out;
}

File: sdk/flutter/dazzle_flutter/lib/src/edge/openai_compatible_client.dart
// Copyright 2026 Ivan Aliaga <ivan.aliaga@urp.edu.pe>
// SPDX-License-Identifier: Apache-2.0
//
// LLMClient implementation over the OpenAI /v1/chat/completions REST
// shape. Works against OpenAI, Groq, HuggingFace Router, vLLM,
// Ollama, llama-server, Together, Mistral — any endpoint that speaks
// the same JSON body + SSE streaming shape.
//
// Pure Dart — no FFI needed, cloud is the bottleneck anyway.

import 'dart:async';
import 'dart:convert';

import 'package:http/http.dart' as http;

import '../agent/llm_client.dart';
import '../agent/message.dart';
import '../agent/tool.dart';

class OpenAICompatibleClient implements LLMClient {
  OpenAICompatibleClient({
    required this.baseURL,
    required String model,
    this.apiKey,
    this.extraHeaders = const {},
    this.temperature,
    this.maxTokens,
    http.Client? httpClient,
  })  : modelId = model,
        _model = model,
        _http = httpClient ?? http.Client();

  final Uri baseURL;
  final String _model;
  final String? apiKey;
  final Map<String, String> extraHeaders;
  final double? temperature;
  final int? maxTokens;
  final http.Client _http;

  @override
  final String modelId;

  @override
  Future<Completion> complete({
    required List<Message> messages,
    List<ToolDeclaration> tools = const [],
  }) async {
    final body = _buildBody(messages, tools, stream: false);
    final resp = await _http.post(
      baseURL.resolve('chat/completions'),
      headers: _headers(),
      body: jsonEncode(body),
    );
    _throwOnHttpError(resp);
    final decoded = jsonDecode(resp.body) as Map<String, dynamic>;
    return _parseCompletionBlocking(decoded);
  }

  @override
  Stream<Delta> stream({
    required List<Message> messages,
    List<ToolDeclaration> tools = const [],
  }) async* {
    final req = http.Request('POST', baseURL.resolve('chat/completions'));
    _headers().forEach((k, v) => req.headers[k] = v);
    req.body = jsonEncode(_buildBody(messages, tools, stream: true));

    final streamed = await _http.send(req);
    if (streamed.statusCode >= 400) {
      final body = await streamed.stream.bytesToString();
      throw StateError('HTTP ${streamed.statusCode}: $body');
    }

    final lines = streamed.stream.transform(utf8.decoder).transform(const LineSplitter());
    final toolIdByIdx = <int, String>{};
    final toolNameById = <String, String>{};

    await for (final rawLine in lines) {
      final line = rawLine.trim();
      if (line.isEmpty || !line.startsWith('data:')) continue;
      final payload = line.substring(5).trim();
      if (payload == '[DONE]') break;

      final json = jsonDecode(payload) as Map<String, dynamic>;
      final choices = json['choices'] as List? ?? const [];
      if (choices.isEmpty) continue;
      final delta = (choices[0] as Map)['delta'] as Map? ?? const {};

      if (delta['content'] case final String text when text.isNotEmpty) {
        yield DeltaText(text);
      }
      if (delta['tool_calls'] case final List calls) {
        for (final callRaw in calls) {
          final call = callRaw as Map<String, dynamic>;
          final idx = call['index'] as int? ?? 0;
          final id = (call['id'] as String?) ?? toolIdByIdx[idx];
          if (id == null) continue;
          toolIdByIdx[idx] = id;

          final fn = call['function'] as Map<String, dynamic>?;
          if (fn != null) {
            if (fn['name'] case final String name) {
              if (!toolNameById.containsKey(id)) {
                toolNameById[id] = name;
                yield DeltaToolCallStart(id: id, name: name);
              }
            }
            if (fn['arguments'] case final String chunk when chunk.isNotEmpty) {
              yield DeltaToolCallArgs(id: id, chunk: chunk);
            }
          }
        }
      }
    }
    yield const DeltaEnd();
  }

  @override
  Future<void> close() async {
    _http.close();
  }

  // MARK: – Helpers

  Map<String, String> _headers() {
    final h = <String, String>{
      'Content-Type': 'application/json',
      if (apiKey != null && apiKey!.isNotEmpty) 'Authorization': 'Bearer $apiKey',
      ...extraHeaders,
    };
    return h;
  }

  Map<String, dynamic> _buildBody(
      List<Message> messages, List<ToolDeclaration> tools,
      {required bool stream}) {
    return {
      'model': _model,
      if (temperature != null) 'temperature': temperature,
      if (maxTokens != null) 'max_tokens': maxTokens,
      'stream': stream,
      'messages': messages.map(_encodeMessage).toList(),
      if (tools.isNotEmpty)
        'tools': tools
            .map((t) => {
                  'type': 'function',
                  'function': {
                    'name': t.name,
                    'description': t.description,
                    'parameters': jsonDecode(t.parameters.serialize()),
                  },
                })
            .toList(),
    };
  }

  Map<String, dynamic> _encodeMessage(Message m) {
    final out = <String, dynamic>{'role': m.role.wire, 'content': m.content};
    if (m.toolCallId != null) out['tool_call_id'] = m.toolCallId;
    if (m.toolCalls.isNotEmpty) {
      out['tool_calls'] = m.toolCalls
          .map((c) => {
                'id': c.id,
                'type': 'function',
                'function': {'name': c.name, 'arguments': c.arguments},
              })
          .toList();
    }
    return out;
  }

  Completion _parseCompletionBlocking(Map<String, dynamic> decoded) {
    final choices = decoded['choices'] as List? ?? const [];
    if (choices.isEmpty) {
      return CompletionText(Message(role: Role.assistant, content: ''));
    }
    final msg = (choices[0] as Map)['message'] as Map? ?? const {};
    final content = (msg['content'] as String?) ?? '';
    final calls = (msg['tool_calls'] as List?) ?? const [];
    if (calls.isNotEmpty) {
      final decoded = calls
          .whereType<Map>()
          .map((c) => ToolCall(
                id: c['id'] as String? ?? '',
                name: (c['function'] as Map?)?['name'] as String? ?? '',
                arguments:
                    (c['function'] as Map?)?['arguments'] as String? ?? '{}',
              ))
          .toList();
      return CompletionToolCalls(Message(
        role: Role.assistant,
        content: content,
        toolCalls: decoded,
      ));
    }
    return CompletionText(Message(role: Role.assistant, content: content));
  }

  void _throwOnHttpError(http.Response resp) {
    if (resp.statusCode >= 400) {
      throw StateError('HTTP ${resp.statusCode} from ${resp.request?.url}: ${resp.body}');
    }
  }
}
